poll/poll.py

```diff
@@ -1236,3 +1276,50 @@ def prepare_data(self):
                         row.append(answers_dict[choice])
                 data[sm.student.id] = row
         return [header_row + questions] + data.values()
+
+    def generate_report_data(self, user_state_iterator, limit_responses=None):
+        """
+        Return a list of student responses to this block in a readable way.
+
+        Arguments:
+            user_state_iterator: iterator over UserStateClient objects.
+                E.g. the result of user_state_client.iter_all_for_block(block_key)
+
+            limit_responses (int|None): maximum number of responses to include.
+                Set to None (default) to include all.
+
+        Returns:
+            each call returns a tuple like:
+            ("username", {
+                           "Question ID": "fun"
+                           "Question": "Are you having fun?"
+                           "Answer": "Yes",
+                           "Answer ID": "1545204793464"
+            })
+        """
+        if limit_responses == 0:
+            return
+
+        answers_dict = dict(self.answers)
+        questions_dict = dict(self.questions)
+        count = 0
+        for user_state in user_state_iterator:
+            # user_state.state={'submissions_count': 1, 'choices': {u'enjoy': u'Y', u'recommend': u'N', u'learn': u'M'}}
+            choices = user_state.state['choices']
+            for q, a in choices.items():
+
+                if limit_responses and count >= limit_responses:
+                    # End the iterator here
+                    return
+
+                question = questions_dict[q]['label']
+                answer = answers_dict[a]
+                report = {
+                    self.ugettext('Question ID'): q,
+                    self.ugettext('Question'): question,
+                    self.ugettext('Answer ID'): a,
+                    self.ugettext('Answer'): answer,
+                    self.ugettext('Submissions count'): user_state.state['submissions_count']
```

Comment:
I don't think Question ID and Answer ID are very useful here since they aren't exposed to the user in any other place.

Reply:
Removed Question ID and Answer ID from both poll and survey reports